File: algorithms/code/leetcode/lc086_partition_list/LC086PartitionList.java
package leetcode.lc086_partition_list;

/**
 * https://leetcode.com/problems/partition-list/
 * #medium
 */
public final class LC086PartitionList {
    public static class ListNode {
        int val;
        ListNode next;

        ListNode(final int x) {
            val = x;
        }
    }

    public ListNode partition(final ListNode headValue, final int x) {
        ListNode head = headValue;
        if (head == null) {
            return null;
        }
        ListNode less = new ListNode(0);
        ListNode more = new ListNode(0);
        ListNode before = less;
        ListNode after = more;
        while (head != null) {
            if (head.val < x) {
                less.next = head;
                less = head;
            }
            if (head.val >= x) {
                more.next = head;
                more = head;
            }
            head = head.next;
        }
        less.next = after.next;
        more.next = null;
        return before.next;
    }
}
package leetcode.lc086_partition_list;

import org.junit.jupiter.api.Test;

import leetcode.lc086_partition_list.LC086PartitionList.ListNode;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;

public final class LC086PartitionListTests {
    private ListNode build(final int[] array) {
        ListNode head = null;
        ListNode tail = null;
        for (int value : array) {
            if (head == null) {
                tail = new ListNode(value);
                head = tail;
            } else {
                tail.next = new ListNode(value);
                tail = tail.next;
            }
        }
        if (tail != null) {
            tail.next = null;
        }
        return head;
    }

    @Test
    public void test143252And3() throws Exception {
        ListNode list = build(new int[] { 1, 4, 3, 2, 5, 2 });
        int[] expected = { 1, 2, 2, 4, 3, 5 };
        ListNode result = new LC086PartitionList().partition(list, 3);
        for (int value : expected) {
            assertEquals(value, result.val);
            result = result.next;
        }
    }

    @Test
    public void testNothing() throws Exception {
        assertNull(new LC086PartitionList().partition(null, 0));
    }
}

File: algorithms/code/leetcode/lc086_partition_list/lc086_partition_list.rs
// https://leetcode.com/problems/partition-list/
// #medium

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

pub struct Solution;

impl Solution {
    pub fn partition(head: Option<Box<ListNode>>, x: i32) -> Option<Box<ListNode>> {
        let mut before = Box::new(ListNode { val: 0, next: None });
        let mut after = Box::new(ListNode { val: 0, next: None });
        let mut less = &mut before;
        let mut more = &mut after;
        let mut head = head;

        while let Some(mut some_head) = head {
            head = some_head.next.take();
            if some_head.val < x {
                less.next = Some(some_head);
                less = less.next.as_mut().unwrap();
            } else {
                more.next = Some(some_head);
                more = more.next.as_mut().unwrap();
            }
        }
        less.next = after.next.take();
        before.next
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn vector_to_list(vector: Vec<i32>) -> Option<Box<ListNode>> {
        let mut list: Option<Box<ListNode>> = None;
        for &value in vector.iter().rev() {
            let node = Box::new(ListNode {
                val: value,
                next: list,
            });
            list = Some(node);
        }
        list
    }

    fn list_to_vector(mut list: Option<Box<ListNode>>) -> Vec<i32> {
        let mut vector = Vec::new();
        while let Some(node) = list {
            vector.push(node.val);
            list = node.next;
        }
        vector
    }

    #[test]
    fn test_143252_and_3() {
        let list = vector_to_list(vec![1, 4, 3, 2, 5, 2]);
        let expected = vec![1, 2, 2, 4, 3, 5];
        let result = Solution::partition(list, 3);
        assert_eq!(list_to_vector(result), expected);
    }

    #[test]
    fn test_nothing() {
        let result = Solution::partition(None, 0);
        assert!(result.is_none());
    }
}
